### DISCUSSION TITLE
Two equal lazy numbers but their double approximations are not equal

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#7044] [Q&A]
Hello,

I have a lazy number (entry 1,1 of the matrix below) which is equal to the lazy number 1 (I compare these two lazy numbers with `==`) but however the double approximation of this number (with relative precision 1e-15) is not equal to 1. How is it possible?

```r
library(lazyNumbers)
# a random 3x3 matrix
set.seed(666L)
M <- matrix(rnorm(9L), nrow = 3L, ncol = 3L)
# the corresponding matrix of lazy numbers
M_lazy <- lazymat(M)
# its inverse
invM_lazy <- lazyInv(M_lazy)
# the product P of the lazy matrix and its inverse
P <- M_lazy %*% invM_lazy
# the identity 3x3 matrix I3
I3 <- diag(3L)
# the double approximation of P is not equal to I3
asDouble(P, prec = 1e-15) == I3 # FALSE for entry 1,1
# however the lazy matrix P is equal to the lazy identity matrix
P == lazymat(I3) # TRUE for all entries
```

**Comment by stla** [comment]
Oops sorry, wanted to post in Q&A.

**Reply by lrineau** [reply]
Can you please show use the code of `asDouble`?

**Reply by stla** [reply]
This is a call from R to this C++ function:

```cpp
Rcpp::NumericVector lvx2nv(lazyVectorXPtr lvx, double prec) {
  lazyVector lv = *(lvx.get());
  const size_t n = lv.size();
  Rcpp::NumericVector nv(n);
  lazyScalar::set_relative_precision_of_to_double(prec);
  for(size_t i = 0; i < n; i++) {
    nv(i) = CGAL::to_double<lazyScalar>(lv[i]);
  }
  return nv;
}
```

**Reply by stla** [reply]
Ah no sorry this is the matricial version:

```cpp
Rcpp::NumericMatrix lmx2nm(lazyMatrixXPtr lmx, double prec) {
  lazyMatrix lm = *(lmx.get());
  const size_t nrow = lm.rows();
  const size_t ncol = lm.cols();
  Rcpp::NumericMatrix nm(nrow, ncol);
  lazyScalar::set_relative_precision_of_to_double(prec);
  for(size_t i = 0; i < nrow; i++) {
    for(size_t j = 0; j < ncol; j++) {
      nm(i, j) = CGAL::to_double<lazyScalar>(lm.coeff(i, j));
    }
  }
  return nm;
}
```

**Comment by stla** [comment]
Aahhh I see (more or less): if I call `asDouble` a second time, then the equality is true!

**Comment by stla** [comment]
No. The equality becomes true if I call `asDouble` after the equality test between the lazy matrices: 

```r
asDouble(P, prec = 1e-15) == I3 # FALSE for entry 1,1
asDouble(P, prec = 1e-15) == I3 # FALSE for entry 1,1
asDouble(P, prec = 1e-15) == I3 # FALSE for entry 1,1
asDouble(P, prec = 1e-15) == I3 # FALSE for entry 1,1
# however the lazy matrix P is equal to the lazy identity matrix
P == lazymat(I3) # TRUE for all entries
asDouble(P, prec = 1e-15) == I3 # TRUE for all entries
```
I guess this is because `==` calls `.exact()`, right?

**Comment by stla** [comment]
I really don't understand that:

```r
library(lazyNumbers)

x <- 1 - lazynb(7)/10
asDouble(x, prec = 1e-15) == 0.3 # TRUE
asDouble(x, prec = 1e-15) == 0.3 # TRUE
asDouble(x, prec = 1e-15) == 0.3 # TRUE
asDouble(x, prec = 1e-15) == 0.3 # TRUE
asDouble(x, prec = 1e-16) == 0.3 # FALSE
asDouble(x, prec = 1e-15) == 0.3 # FALSE ???
```

**Reply by stla** [reply]
> I really don't understand that:
>  ...

Ah, that's because the call with `1e-16` performs a refinement?

Anyway, now I apply `.exact` before doing the double approximation, and nothing strange happens in this way.

**Reply by lrineau** [reply]
Sorry, the CGAL team has not been very responsive last week.

Calling `.exact()` is indeed the only way to get a reproducible double approximation of an exact number. Otherwise, you can have all the surprises you painfully discovered during your attempts.

I hope your attempt to use CGAL lazy numbers is now a success. Let us know otherwise.

**Reply by stla** [reply]
Yep, my package is on CRAN now :-) No problem so far but they haven't check with all the platforms yet.
